### PR TITLE
NooBaa Operator - Phase 1 (DRAFT)

### PR DESCRIPTION
**Pull request by guymguym**
**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../CONTRIBUTING.md#comments)

// while WIP
[skip ci]

### CODE REVIEW

File: pkg/operator/noobaa/operator.go [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+}
+
+// CreateCRD uses APIExtClient to create the CRD if it doesn't exist.
+// Returns true if created/exists, false to retry, error on unrecoverable error.
+func (o *Operator) CreateCRD(crd *ext_v1.CustomResourceDefinition) (bool, error) {
+	crdClient := o.APIExtClient.ApiextensionsV1beta1().CustomResourceDefinitions()
+	crdReply, err := crdClient.Get(crd.GetName(), meta_v1.GetOptions{})
+	if errors.IsNotFound(err) {
+		_, err = crdClient.Create(crd)
```
Comment (guymguym) [inline pkg/operator/noobaa/operator.go]: @travisn I saw you wrote that rook operators don't have permission to create CRDs.
So far I was running the operator with the permissions, but I can make it optional so it will try to create once or so.
What do you think?

Comment (travisn) [reply pkg/operator/noobaa/operator.go]: Long ago the rook-ceph operator created all the resources automatically it needed... the CRDs, service accounts, etc. This required the operator to have full privileges for the lifetime of the operator, which is indefinitely. After a security review, we moved to minimal privileges where we define only the privileges absolutely necessary with the RBAC. This ensures that the long-running operator doesn't have admin privs. The admin can use their higher privs once to create the CRDs, service accounts, etc, rather than requiring the operator to have them. It isn't as clean of a UX since all the yaml must be creaed separately by the admin, but it's a much better security approach. See [common.yaml](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/common.yaml) for rook-ceph's RBAC, though it will probably be much more complex than what Noobaa needs.

Comment (guymguym) [reply pkg/operator/noobaa/operator.go]: OK thanks, will follow the same.

File: pkg/operator/noobaa/system_state.go [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+	if sys.Spec.Image == "" {
+		sys.Spec.Image = ContainerImage
+	}
+
+	makeInt32Ptr := func(x int32) *int32 { return &x }
+
+	makeObjectMeta := func(name string) meta_v1.ObjectMeta {
+		ownerIsController := true
+		owner := meta_v1.OwnerReference{
+			APIVersion: sys.APIVersion,
```
Comment (travisn) [inline pkg/operator/noobaa/system_state.go]: There was a bug in the owner references creation that you'll want to pick up on this line with `APIVersion`. See #3350.

File: cmd/rook/noobaa/noobaa.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+// Cmd is the parent command for noobaa operator and tools
+var Cmd = &cobra.Command{
+	Use:   "noobaa",
+	Short: "Main command for noobaa operator.",
+	Long:  "See https://github.com/noobaa/noobaa-core and https://github.com/rook/rook",
+}
+
+// ASCIILogo is noobaa's logo ascii art
+const ASCIILogo = `
```
Comment (galexrt) [inline cmd/rook/noobaa/noobaa.go]: Normally the `rook` command is not run directly, so why print the logo?

Personally speaking, printing a logo in the logs just means unnecessary lines of logs (e.g., to be shipped to Elasticsearch).
Though I think I know where you are coming from with this.

File: pkg/operator/noobaa/nbapi/rpc.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	method := body.Method
+	if body.AuthToken == "" {
+		body.AuthToken = c.AuthToken
+	}
+	address := c.Router.GetAddress(api)
+	logger.Infof("NBAPI: %s.%s - RPC Call to %s: %+v", api, method, address, req)
+
+	reqBytes, err := json.Marshal(body)
+	if err != nil {
+		logger.Errorf("NBAPI: ❌ %s.%s - Encoding request failed: %s", api, method, err)
```
Comment (jbw976) [inline pkg/operator/noobaa/nbapi/rpc.go]: are these emoji chars in the log error strings?  i'm not sure if that's part of the string or if github is rendering this diff weird haha

does having non ascii chars in log strings potentially break any logging systems that may be collecting/aggregating/viewing these logs?

File: pkg/operator/noobaa/system_state.go [hunk 5]
```diff
@@ -0,0 +1,698 @@
...
+								{ContainerPort: 8446},
+								{ContainerPort: 60100},
+							},
+							VolumeMounts: []core_v1.VolumeMount{
+								{MountPath: "/data", Name: datadirPVCName},
+								{MountPath: "/log", Name: logdirPVCName},
+							},
+							Env: []core_v1.EnvVar{
+								{Name: "CONTAINER_PLATFORM", Value: "KUBERNETES"},
+								// - name: CREATE_SYS_NAME
```
Comment (jbw976) [inline pkg/operator/noobaa/system_state.go]: are all these commented out pod spec env vars needed? they don't seem to be informing the code in any way currently

File: pkg/operator/noobaa/system_state.go (hunk 2 above)
Comment (guymguym) [reply pkg/operator/noobaa/system_state.go]: I am copying the APIVersion from the NooBaaSystem object, so I think it's good. Thanks.

File: pkg/apis/noobaa.rook.io/v1alpha1/types.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+// ***************************************************************************
+// IMPORTANT FOR CODE GENERATION
+// If the types in this file are updated, you will need to run
+// `make codegen` to generate the new types under the client/clientset folder.
+// ***************************************************************************
+
+import (
+	"reflect"
+
+	core_v1 "k8s.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (guymguym) [reply pkg/apis/noobaa.rook.io/v1alpha1/types.go]: Less readable imo, but I prefer to follow it if this is a common practice.
Resolving, thanks.

File: images/noobaa/Makefile [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+TEMP := $(shell mktemp -d)
+
+# ====================================================================================
+# Build Rook
+
+do.build:
+	@echo === docker build $(NOOBAA_IMAGE)
+	@cp Dockerfile $(TEMP)
+	@cp $(OUTPUT_DIR)/bin/linux_$(GOARCH)/rook $(TEMP)/rook-noobaa-operator
```
[the remark this answers is not in the document]
Comment (guymguym) [reply images/noobaa/Makefile]: I think of the rook binary as a collection of tools, so here in the build process for the noobaa operator image I thought it made sense that the process name will be descriptive as well. Maybe its because I'm used to looking at the linux level. In any way the process name inside the container is not really important.

**TLDR: Changed. Resolving.**

File: images/noobaa/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM alpine:3.9
+
+WORKDIR /rook
+COPY . ./
+
+ENTRYPOINT ["/rook/rook-noobaa-operator"]
```
[the remark this answers is not in the document]
Comment (guymguym) [reply images/noobaa/Dockerfile]: Oh yes, it does take "noobaa" and "operator" as args, it's in the pod spec.
Anyway I renamed the binary to rook as discussed in your other comment.
Resolving.

File: cmd/rook/noobaa/noobaa.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	initSubCommand(&cobra.Command{
+		Hidden: true,
+		Use:    "status",
+		Short:  "Read system status",
+		Run: func(cmd *cobra.Command, args []string) {
+			logger.Infof("TODO: %s ...", cmd.Use)
+		},
+	})
+
+	initSubCommand(&cobra.Command{
```
[the remark this answers is not in the document]
Comment (guymguym) [reply cmd/rook/noobaa/noobaa.go]: I want noobaa users to run rook as the CLI tool to ease a user day1 and day2 operations, like kubectl or [velero cli](https://velero.io/docs/v1.0.0/install-overview/).
So these actions here are meant to setup the operator for a user, and list/create/delete CRs that the operator will act upon, and show user friendly statuses.
I think that docs only is less attractive to users on the dev* spectrum, than a discoverable command line action menu like velero/kubectl/etc.

@travisn @jbw976 What do you think?

NooBaa CLI would include for example:
```
rook
  noobaa
    - install
    - uninstall
    - operator
        - run
        - restart
    - system
        - list
        - create
        - delete
        - status
     - bucket
        - list
        - create
        - delete
     - backing-store
        - list
...
```

File: images/noobaa/Makefile [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+	@cp ../../cluster/examples/kubernetes/noobaa/* $(TEMP)
```
[the remark this answers is not in the document]
Comment (guymguym) [reply images/noobaa/Makefile]: I was using it for the operator to create the CRD's but will change so that the operator won't have these permissions and remove from the image.

File: pkg/operator/noobaa/nbapi/api.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
+/*
```
[the remark this answers is not in the document]
Comment (guymguym) [reply pkg/operator/noobaa/nbapi/api.go]: Changing. Thanks

File: README.md [hunk 12]
```diff
@@ -75,6 +75,7 @@ Cassandra      | [Cassandra](http://cassandra.apache.org/) is a highly available
...
 Minio          | [Minio](https://www.minio.io/) is a high performance distributed object storage server, designed for large-scale private cloud infrastructure.                                                                                                                                                                         | minio.rook.io/v1alpha1       | Alpha
 NFS            | [Network File System (NFS)](https://github.com/nfs-ganesha/nfs-ganesha/wiki) allows remote hosts to mount file systems over a network and interact with those file systems as though they are mounted locally.                                                                                                         | nfs.rook.io/v1alpha1         | Alpha
+NooBaa         | [NooBaa](https://github.com/noobaa/noobaa-core) is an object data service for hybrid and multi cloud environments. | noobaa.rook.io/v1alpha1 | Alpha
```
[the remark this answers is not in the document]
Comment (guymguym) [reply README.md]: I just followed previous rows, and the entire table renders fine since outer pipes are optional.
See [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables).
Resolving, thanks.

File: cluster/examples/kubernetes/noobaa/noobaa-backing-store-crd.yaml [hunk 13]
```diff
@@ -1,14 +1,46 @@
+###############################################################################
+#
+# CRD: NooBaaBackingStore
+#
+# WAIT !!! No need to create this manually, the operator does it.
```
[the remark this answers is not in the document]
Comment (guymguym) [reply cluster/examples/kubernetes/noobaa/noobaa-backing-store-crd.yaml]: Thanks! I do have a question on the suggested security model - when the operator is responsible for a certain resource, say a CRD or a Role, it can update it on operator upgrade to match the operator version. How is this supposed to work in this secure model? What if the operator relies on a certain role to be applied in the new version? What if the CRD validation changes between versions?

File: cluster/examples/kubernetes/noobaa/noobaa-backing-store-crd.yaml [hunk 14]
```diff
@@ -1,14 +1,46 @@
...
+#
+###############################################################################
+
 apiVersion: apiextensions.k8s.io/v1beta1
 kind: CustomResourceDefinition
 metadata:
   name: noobaabackingstores.noobaa.rook.io
 spec:
+
```
[the remark this answers is not in the document]
Comment (guymguym) [reply cluster/examples/kubernetes/noobaa/noobaa-backing-store-crd.yaml]: Okie dokie. Changing.

File: cluster/examples/kubernetes/noobaa/noobaa-operator.yaml [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+    resources: ["deployment", "replicasets", "statefulsets", "daemonsets"]
+    verbs:     ["get", "list", "watch", "create", "update", "patch", "delete"]
+  # rbac
+  - apiGroups: ["rbac.authorization.k8s.io"]
+    resources: ["roles", "rolebindings"]
+    verbs:     ["get", "list", "watch", "create", "update", "patch", "delete"]
+  # crd
+  - apiGroups: ["apiextensions.k8s.io"]
+    resources: ["customresourcedefinitions"]
+    verbs:     ["get", "list", "watch", "create", "update", "patch", "delete"]
```
[the remark this answers is not in the document]
Comment (guymguym) [reply cluster/examples/kubernetes/noobaa/noobaa-operator.yaml]: I can limit it to just the resource names that noobaa operator is responsible to - would that be considered safe?

In general - can you describe the rational behind the security risk in giving permissions to CRD's to an operator?

File: cmd/rook/noobaa/noobaa.go [hunk 16]
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2018 The Rook Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (guymguym) [reply cmd/rook/noobaa/noobaa.go]: Changing.

File: cluster/examples/kubernetes/noobaa/noobaa-operator.yaml [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+  labels:
+    app: rook-noobaa-operator
+rules:
+  # read-only
+  - apiGroups: [""]
+    resources: ["namespaces", "nodes"]
+    verbs:     ["get", "list", "watch"]
+  # core
+  - apiGroups: [""]
+    resources: ["pods", "services", "secrets", "configmaps", "serviceaccounts", "persistentvolumeclaims"]
```
[the remark this answers is not in the document]
Comment (guymguym) [reply cluster/examples/kubernetes/noobaa/noobaa-operator.yaml]: Just service accounts? or you mean also the rbac roles and bindings?

File: cmd/rook/noobaa/noobaa.go (hunk 3 above)
Comment (guymguym) [reply cmd/rook/noobaa/noobaa.go]: It is printed once on operator startup so I don't think it matters to anyone.
The reason to print a logo is ... well there is no reason, it is just beautiful.
I prefer to keep it for now, but will certainly remove it if others will complain.
Thanks! Resolving.

File: Documentation/development-flow.md [hunk 18]
```diff
@@ -277,3 +277,48 @@ git push origin HEAD
...
+
+### Building locally
+
+Building a single rook binary for all operators:
+
+```
+make GO_STATIC_PACKAGES=github.com/rook/rook/cmd/rook go.build
+```
+
+Note: the binary output location is `_output/bin/linux_amd64/rook` on linux, and `_output/bin/darwin_amd64/rook` on mac.
```
[the remark this answers is not in the document]
Comment (guymguym) [reply Documentation/development-flow.md]: Great! Changed.

File: build/makelib/common.mk [hunk 19]
```diff
@@ -98,8 +98,12 @@ ifeq ($(BUILD_REGISTRY),build-)
 $(error Failed to get unique ID for host+dir. Check that '$(SHA256CMD)' functions or override SHA256CMD)
 endif
 
+# This is a neat little target that prints any variable value from the Makefile
+# Usage: make echo.IMAGES echo.PLATFORM
+echo.%: ; @echo $* = $($*)
```
[the remark this answers is not in the document]
Comment (guymguym) [reply build/makelib/common.mk]: Glad you find it useful!

File: pkg/operator/noobaa/nbapi/rpc.go (hunk 4 above)
Comment (guymguym) [reply pkg/operator/noobaa/nbapi/rpc.go]: Yes, these are unicode characters to make the logs more human readable.
I certainly hope it will not break machine readability because unicode chars might be used in resource names given by users too, right?
So I think to keep it for now and wait for first pushback.
Thanks

File: pkg/apis/noobaa.rook.io/v1alpha1/types.go [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type NooBaaSystemList struct {
+	meta_v1.TypeMeta `json:",inline"`
+	meta_v1.ListMeta `json:"metadata"`
+	Items            []NooBaaSystem `json:"items"`
+}
+
+// SystemSpec is the spec of a noobaa system
+type SystemSpec struct {
+	// ServerImage (optional) overrides the default image for server pods
```
[the remark this answers is not in the document]
Comment (guymguym) [reply pkg/apis/noobaa.rook.io/v1alpha1/types.go]: Oops, Fixed. Resolving.

File: pkg/operator/noobaa/system_state.go (hunk 5 above)
Comment (guymguym) [reply pkg/operator/noobaa/system_state.go]: You're right. Removed it.

File: cmd/rook/noobaa/noobaa.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (travisn) [reply cmd/rook/noobaa/noobaa.go]: Having an operator and CRDs is a new mindset that is a different way of doing things. The approach includes that kubectl is the only command line tool that will effectively do anything with the operator or CRDs such as create, delete, list, update. Here are a few example commands:
- Create the operator: `kubectl create -f operator.yaml`
- Create a bucket: `kubectl create -f bucket.yaml`
- Delete a bucket: `kubectl delete noobaabucket mybucket` 
- List buckets: `kubectl get noobaabucket`
- Get properties for a specific bucket: `kubectl get noobaabucket mybucket`
   - Assumes that interesting properties for the bucket are added by the operator to the `status` on the CR, then the `additionalPrinterColumns` is included with the CRD declaration.

So the admin really is only expected to interact with noobaa through `kubectl`. Other commands are purely internal to the operator. An exception we have to this for rook-ceph is that we have a toolbox where you can run any ceph command imaginable. However, that's not for common scenarios. It's only for advanced troubleshooting.

And per other comments, the operator shouldn't have privileges to install itself by creating service accounts, RBAC rules, etc. The operator should have only the privileges needed to handle the CRDs.

File: cluster/examples/kubernetes/noobaa/noobaa-backing-store-crd.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (galexrt) [reply cluster/examples/kubernetes/noobaa/noobaa-backing-store-crd.yaml]: @guymguym The person that manages the operator is responsible for updating the RBAC, CRDs and other "parts" during upgrades, so the operator just runs a reconciliation loop every time it starts (e.g., as Ceph does when operators starts, a new disk is added).

If there are breaking changes to a CRD, the operator will take of migrating them to, e.g., a new CRD version (e.g., `v1alpha1` -> `v1alpha2`).
Searching for "migration", e.g., https://github.com/rook/rook/blob/818ad9c7ca4d352a9bed202e619d47b05f1a4dab/pkg/operator/ceph/cluster/migration.go, should give you an insight to how Ceph operator does it right now.

File: cluster/examples/kubernetes/noobaa/noobaa-bucket-class-crd.yaml [hunk 21]
```diff
@@ -1,14 +1,41 @@
...
+# WAIT !!! No need to create this manually, the operator does it.
+#
+###############################################################################
 apiVersion: apiextensions.k8s.io/v1beta1
 kind: CustomResourceDefinition
 metadata:
   name: noobaabucketclasses.noobaa.rook.io
 spec:
+  scope: Namespaced
+  group: noobaa.rook.io
```
Comment (jeffvance) [inline cluster/examples/kubernetes/noobaa/noobaa-bucket-class-crd.yaml]: Is this namespaced resource required only for noobaa, or is it meant to be an addition to general bucket provisioning?

**Comment by jeffvance** [comment]
This PR looks thorough (other than missing tests as pointed out). In skimming the code and thinking about this PR vs. pr #3067 and the bucket lib, I see asymmetry. This pr provides all the code for noobaa clientset, listers, informers, versioned types, etc, as done by ceph, nfs, minio, etc. 3067 does not. It imports a lib which handles these tasks in a more generic manner. I'm trying to reconcile the differences in my head but stalling. For now I leave this as an intro comment...

**Comment by jeffvance** [comment]
@guymguym Should this PR be closed per previous slack comment?
_this is a stale PR, we are not developing it further. You should be looking on the noobaa-operator repo_
